lib/plugins/authpdo/auth.php

```diff
@@ -0,0 +1,738 @@
+            )
+        );
+
+        $this->success = true;
+    }
+
+    /**
+     * Check user+password
+     *
+     * @param   string $user the user name
+     * @param   string $pass the clear text password
+     * @return  bool
+     */
+    public function checkPass($user, $pass) {
+
+        $data = $this->_selectUser($user);
+        if($data == false) return false;
+
+        if(isset($data['hash'])) {
+            // hashed password
+            $passhash = new PassHash();
+            return $passhash->verify_hash($pass, $data['hash']);
+        } else {
+            // clear text password in the database O_o
+            return ($pass == $data['clear']);
+        }
+    }
+
+    /**
+     * Return user info
+     *
+     * Returns info about the given user needs to contain
+     * at least these fields:
+     *
+     * name string  full name of the user
+     * mail string  email addres of the user
+     * grps array   list of groups the user is in
+     *
+     * @param   string $user the user name
+     * @param   bool $requireGroups whether or not the returned data must include groups
+     * @return array|bool containing user data or false
+     */
+    public function getUserData($user, $requireGroups = true) {
+        $data = $this->_selectUser($user);
+        if($data == false) return false;
+
+        if(isset($data['hash'])) unset($data['hash']);
+        if(isset($data['clean'])) unset($data['clean']);
+
+        if($requireGroups) {
+            $data['grps'] = $this->_selectUserGroups($data);
+            if($data['grps'] === false) return false;
+        }
+
+        return $data;
+    }
+
+    /**
+     * Create a new User [implement only where required/possible]
+     *
+     * Returns false if the user already exists, null when an error
+     * occurred and true if everything went well.
+     *
+     * The new user HAS TO be added to the default group by this
+     * function!
+     *
+     * Set addUser capability when implemented
+     *
+     * @param  string $user
+     * @param  string $clear
+     * @param  string $name
+     * @param  string $mail
+     * @param  null|array $grps
+     * @return bool|null
+     */
+    public function createUser($user, $clear, $name, $mail, $grps = null) {
+        global $conf;
+
+        if(($info = $this->getUserData($user, false)) !== false) {
+            msg($this->getLang('userexists'), -1);
+            return false; // user already exists
+        }
+
+        // prepare data
+        if($grps == null) $grps = array();
+        array_unshift($grps, $conf['defaultgroup']);
+        $grps = array_unique($grps);
+        $hash = auth_cryptPassword($clear);
+        $userdata = compact('user', 'clear', 'hash', 'name', 'mail');
+
+        // action protected by transaction
+        $this->pdo->beginTransaction();
+        {
+            // insert the user
+            $ok = $this->_query($this->getConf('insert-user'), $userdata);
+            if($ok === false) goto FAIL;
```

Comment:
this shouldn't use goto for error handling. Db handling is already partly in separate functions, the error handling should be handled this way too. Another approach could be to separate:

- Writing the user to the DB
- preparing the data for for storage
- handling errors